### PR TITLE
Limit concurrent HTTP/2 Streams per connection

### PR DESCRIPTION
**Pull request by Tratcher**
#2815 Defaults to 100 which is similar to Nginx, Apache, and IIS/Http.Sys.

I needed to finish implementing the infrastructure for sending settings to the client.

I also experimented with the BinaryConverter APIs for #2773, they worked great.

### CODE REVIEW

File: src/Kestrel.Core/Internal/Http2/Http2Frame.Settings.cs [hunk 1]
```diff
@@ -1,42 +1,67 @@
...
+            var id = (Http2SettingsParameter)BinaryPrimitives.ReadUInt16BigEndian(payload);
+            var value = BinaryPrimitives.ReadUInt32BigEndian(payload.Slice(2));
+
+            return new Http2PeerSetting(id, value);
+        }
 
-            Length = 6 * settingCount;
+        public void PrepareSettings(Http2SettingsFrameFlags flags, IList<Http2PeerSetting> settings = null)
+        {
+            var settingCount = settings?.Count() ?? 0;
```
Comment (campersau) [inline src/Kestrel.Core/Internal/Http2/Http2Frame.Settings.cs]: Can be `settings?.Count ?? 0` now.

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Frame.Settings.cs]: Yes, for the Settings Ack code path no settings are sent.

File: src/Kestrel.Core/KestrelServerLimits.cs [hunk 2]
```diff
@@ -251,6 +252,25 @@ public TimeSpan RequestHeadersTimeout
             }
         }
 
+        /// <summary>
+        /// Limits the number of concurrent request streams per HTTP/2 connection. Excess streams will be refused.
+        /// <para>
+        /// Defaults to 100
+        /// </para>
+        /// </summary>
+        public int MaxStreamsPerConnection
```
Comment (halter73) [inline src/Kestrel.Core/KestrelServerLimits.cs]: What do you think about making an Http2Limits class/property and adding this to that?

File: src/Kestrel.Core/KestrelServerLimits.cs [hunk 3]
```diff
@@ -251,6 +252,25 @@ public TimeSpan RequestHeadersTimeout
...
+        {
+            get => _maxStreamsPerConnection;
+            set
+            {
+                if (value < 0)
```
Comment (halter73) [inline src/Kestrel.Core/KestrelServerLimits.cs]: `<=`

Comment (Tratcher) [reply src/Kestrel.Core/KestrelServerLimits.cs]: 0 is technically allowed. You can have the endpoint up for pings but not accept requests.

File: src/Kestrel.Core/KestrelServerLimits.cs (hunk 2 above)
Comment (Tratcher) [reply src/Kestrel.Core/KestrelServerLimits.cs]: No, there won't be that many. Also most of the existing limits apply to Http/2.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 4]
```diff
@@ -838,6 +841,11 @@ private void StartStream<TContext>(IHttpApplication<TContext> application)
                 throw new Http2StreamErrorException(_currentHeadersStream.StreamId, CoreStrings.Http2ErrorMissingMandatoryPseudoHeaderFields, Http2ErrorCode.PROTOCOL_ERROR);
             }
 
+            if (_streams.Count >= _serverSettings.MaxConcurrentStreams)
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Is there a potential race here? Between checking the _streams.Count and adding the _currentHeadersStream to _streams?

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: We're under the state lock, so there shouldn't be.

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: We also only ever process one frame at a time.

File: src/Kestrel.Core/Internal/Http2/Http2PeerSettings.cs [hunk 5]
```diff
@@ -91,16 +85,42 @@ public void ParseFrame(Http2Frame frame)
...
+        internal IList<Http2PeerSetting> GetNonProtocolDefaults()
         {
-            yield return new Http2PeerSetting(Http2SettingsParameter.SETTINGS_HEADER_TABLE_SIZE, HeaderTableSize);
-            yield return new Http2PeerSetting(Http2SettingsParameter.SETTINGS_ENABLE_PUSH, EnablePush ? 1u : 0);
-            yield return new Http2PeerSetting(Http2SettingsParameter.SETTINGS_MAX_CONCURRENT_STREAMS, MaxConcurrentStreams);
-            yield return new Http2PeerSetting(Http2SettingsParameter.SETTINGS_INITIAL_WINDOW_SIZE, InitialWindowSize);
-            yield return new Http2PeerSetting(Http2SettingsParameter.SETTINGS_MAX_FRAME_SIZE, MaxFrameSize);
-            yield return new Http2PeerSetting(Http2SettingsParameter.SETTINGS_MAX_HEADER_LIST_SIZE, MaxHeaderListSize);
-        }
+            var list = new List<Http2PeerSetting>(0);
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2PeerSettings.cs]: Given that `MaxConcurrentStreams != DefaultMaxConcurrentStreams` by default (ironic no?), shouldn't we at least pass 1 into the List ctor? Also, what's wrong with IEnumerable and yield return?

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs (hunk 4 above)
Comment (JunTaoLuo) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Oh yea, I added that state lock. I just forgot to check the callsite.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 6]
```diff
@@ -620,7 +623,7 @@ private Task ProcessSettingsFrameAsync()
                 // ParseFrame will not parse an InitialWindowSize > int.MaxValue.
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Nit: fix comment.

File: src/Kestrel.Core/KestrelServerLimits.cs (hunk 2 above)
Comment (JunTaoLuo) [reply src/Kestrel.Core/KestrelServerLimits.cs]: But for Http/2 specific limits, maybe we should prefix them to something like `HTTP2MaxStreamsPerConnection`?

Comment (halter73) [reply src/Kestrel.Core/KestrelServerLimits.cs]: What all do we expect? I can already think of these potentially:

- MaxStreamsPerConnection
- TotalMaxStreams
- ConnectionWindowSize
- StreamWindowSize
- WindowUpdateInterval
- MaxFrameSize

Even if we only do a couple of these, I like having non-HTTP/1.x-related limits separated. In the future, we could add even more related to header table size and possibly other things we haven't even considered yet.

File: src/Kestrel.Core/KestrelServerLimits.cs (hunk 3 above)
Comment (halter73) [reply src/Kestrel.Core/KestrelServerLimits.cs]: Why though? Even if if allowed by protocol, we don't need to allow a pit of failure.

File: src/Kestrel.Core/KestrelServerLimits.cs (hunk 2 above)
Comment (Tratcher) [reply src/Kestrel.Core/KestrelServerLimits.cs]: I asked @shirhatti and he wants to keep it as a flat list.

**Comment by Tratcher** [comment]
Minor updates

**Pull request merged**